python/lsst/dax/obscore/script/obscore_update_table.py

```diff
@@ -0,0 +1,132 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ["obscore_update_table"]
+
+import logging
+import re
+import time
+from collections.abc import Iterator
+from typing import Any
+
+from lsst.daf.butler import Butler, CollectionType, DatasetType
+from lsst.daf.butler.registries.sql import SqlRegistry
+from lsst.daf.butler.registry.interfaces import CollectionRecord
+from lsst.daf.butler.registry.obscore import (
+    ConfigCollectionType,
+    ObsCoreLiveTableManager,
+    ObsCoreManagerConfig,
+)
+from lsst.utils import iteration
+
+_LOG = logging.getLogger(__name__)
+
+
+def obscore_update_table(
+    repo: str,
+    dry_run: bool,
+) -> None:
+    """Update obscore table with the records that are missing, typically
+    used after adding obscore support to existing repository.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI to the butler repository.
+    dry_run : `bool`
+        If `True` then print the records that will be added, but do not
+        actually add them.
+    """
+
+    butler = Butler(repo, writeable=True)
+
+    # There is no client API for updating obscore table, so we need to access
+    # internals of the Registry and obscore manager.
+    registry = butler.registry
+    assert isinstance(registry, SqlRegistry), "Registry must be SqlRegistry"
+    manager = registry._managers.obscore
+    assert manager is not None, "Registry is not configured for obscore support"
+    assert isinstance(manager, ObsCoreLiveTableManager), f"Unexpected type of obscore manager {type(manager)}"
+    config = manager.config
+
+    for collection_record, dataset_type in _collections(registry, config):
+        start_time = time.time()
+        refs = registry.queryDatasets(dataset_type, collections=collection_record.name).expanded()
+        if dry_run:
+            for ref in refs:
+                _LOG.info("Will be adding dataset %s", ref)
+        else:
+            count = 0
+            if collection_record.type is CollectionType.RUN:
+                # Limit record number in single insert.
+                for refs_chunk in iteration.chunk_iterable(refs):
+                    count += manager.add_datasets(refs_chunk)
+            elif collection_record.type is CollectionType.TAGGED:
+                for refs_chunk in iteration.chunk_iterable(refs):
+                    count += manager.associate(refs_chunk, collection_record)
+            else:
+                raise ValueError(f"Unexpected collection type: {collection_record.type}")
+            end_time = time.time()
+            _LOG.info(
```

Comment:
Consider using `lsst.utils.timer.time_this` which lets you time a block and issue a log message on the logger of your choice with the specified level. Saves you having to do the `time.time()` thing (granted that this might be complicated if you don't want a timer to be issued for the dry_run branch).

Reply:
I think I like this message better:
- one issue is that I can only construct message itself at the end of the block,
- I only need to know how long it took, extra noise about memory use in the message is not really needed here.